### DISCUSSION TITLE
SPI chip select pin

### DISCUSSION THREAD

**Discussion by thrystan** [tagyoureit/nodejs-poolController#763] [REM]
How do I tell REM which pin is being used for the SPI chip select pin? In the adafruit sample code I have been using there is a parameter to specify the cs pin, but I don't see an option to do that in REM.

**Comment by rstrouse** [comment]
Raspberry Pi has two SPI buses built into the os.  Those are set by the boot config but not all pins can be assigned to a CS for SPI.

**Reply by thrystan** [reply]
I'm using spi bus 0 with the CS0 pin, with the mcp3008 chip and I'm not getting any data from the pressure transducer I have connected to it.

**Reply by rstrouse** [reply]
Are you reading from REM?

**Reply by thrystan** [reply]
I'm trying to, I can read it from my python script, but not from REM. 
Speaking of which. I was adding in my 1-wire air temp sensor, and I hit the restart button, and now I'm getting the following error.

REM > error: ORP Cannot read properties of undefined (reading 'split')                                
REM > info: Device Atlas EZO-ORP initialized for i2c bus #1 address 98                                
REM > info: Device Atlas EZO-pH initialized for i2c bus #1 address 99                                 
REM > info: Device Multi-Relay Board initialized for i2c bus #1 address 19                            
REM > info: i2c Bus #1 Initialized                                                                    
REM > error: Cannot read properties of undefined (reading 'toLowerCase')                              
REM > /home/pi/relayEquipmentManager/spi-adc/SpiAdcBus.ts:133                                         
REM >         switch (this.device.input.toLowerCase()) {                                              
REM >                                   ^                                                             
REM > TypeError: Cannot read properties of undefined (reading 'toLowerCase')                          
REM >     at SpiAdcChannel.convertValue (/home/pi/relayEquipmentManager/spi-adc/SpiAdcBus.ts:133:35)  
REM >     at /home/pi/relayEquipmentManager/spi-adc/SpiAdcBus.ts:212:52

**Reply by rstrouse** [reply]
Can I get a copy of your `~/relayEquipmentManager/data/controllerConfig.json`?

**Reply by thrystan** [reply]
{
  "controllerType": "raspi",
  "gpio": {
    "pins": [
      {
        "id": 36,
        "headerId": 1,
        "isInverted": false,
        "direction": "input",
        "triggers": [],
        "initialState": "last",
        "sequenceOnDelay": 0,
        "sequenceOffDelay": 0,
        "name": "Pump on",
        "feeds": [],
        "isActive": true,
        "debounceTimeout": 0,
        "state": "off"
      },
      {
        "id": 16,
        "headerId": 1,
        "isInverted": false,
        "direction": "input",
        "triggers": [],
        "initialState": "last",
        "sequenceOnDelay": 0,
        "sequenceOffDelay": 0,
        "name": "Flow Meter",
        "feeds": [],
        "isActive": true,
        "debounceTimeout": 0,
        "state": "on"
      },
      {
        "id": 11,
        "headerId": 1,
        "isInverted": false,
        "direction": "output",
        "triggers": [],
        "initialState": "on",
        "sequenceOnDelay": 0,
        "sequenceOffDelay": 0,
        "name": "Pin #1-11",
        "feeds": [],
        "isActive": false,
        "debounceTimeout": 0,
        "state": "off"
      },
      {
        "id": 12,
        "headerId": 1,
        "isInverted": false,
        "direction": "output",
        "triggers": [],
        "initialState": "off",
        "sequenceOnDelay": 100,
        "sequenceOffDelay": 100,
        "name": "pin 18",
        "feeds": [],
        "isActive": true,
        "debounceTimeout": 0,
        "state": "off"
      },
      {
        "id": 15,
        "headerId": 1,
        "isInverted": false,
        "direction": "output",
        "triggers": [],
        "initialState": "off",
        "sequenceOnDelay": 0,
        "sequenceOffDelay": 0,
        "name": "pin22",
        "feeds": [],
        "isActive": true,
        "debounceTimeout": 0,
        "state": "off"
      },
      {
        "id": 22,
        "headerId": 1,
        "isInverted": false,
        "direction": "output",
        "triggers": [],
        "initialState": "last",
        "sequenceOnDelay": 0,
        "sequenceOffDelay": 0,
        "name": "Pool Light",
        "feeds": [],
        "isActive": true,
        "debounceTimeout": 0,
        "state": "off"
      }
    ],
    "exported": [
      16,
      23,
      18,
      22,
      25
    ]
  },
  "spi0": {
    "isActive": true,
    "busNumber": 0,
    "channels": [
      {
        "id": 1,
        "isActive": false,
        "feeds": [],
        "sampling": 1
      },
      {
        "id": 2,
        "isActive": true,
        "feeds": [],
        "deviceId": 202,
        "options": {
          "0": {
            "0": "",
            "1": true,
            "2": {
              "0": null,
              "1": null
            }
          },
          "1": {
            "0": {
              "0": {
                "0": 0.5,
                "1": 25
              }
            },
            "1": {
              "1": false,
              "2": 0,
              "3": 0
            }
          }
        },
        "sampling": 1
      },
      {
        "id": 3,
        "isActive": false,
        "feeds": [],
        "sampling": 1
      },
      {
        "id": 4,
        "isActive": false,
        "feeds": [],
        "sampling": 1
      },
      {
        "id": 5,
        "isActive": false,
        "feeds": [],
        "sampling": 1
      },
      {
        "id": 6,
        "isActive": false,
        "feeds": [],
        "sampling": 1
      },
      {
        "id": 7,
        "isActive": false,
        "feeds": [],
        "sampling": 1
      },
      {
        "id": 8,
        "isActive": false,
        "feeds": [],
        "sampling": 1
      }
    ],
    "referenceVoltage": 3.3,
    "spiClock": 1350,
    "adcChipType": 1
  },
  "spi1": {
    "isActive": false,
    "busNumber": 1,
    "channels": [],
    "referenceVoltage": 3.3,
    "spiClock": 1000
  },
  "i2c": {
    "isActive": false,
    "buses": [
      {
        "id": 1,
        "isActive": true,
        "devices": [
          {
            "id": 1,
            "values": {
              "orp": 690.1
            },
            "options": {
              "isProtocolLocked": false,
              "ledEnabled": false,
              "readInterval": 99000,
              "calibrationMode": true,
              "calibration": {
                "len": 2,
                "total": 14,
                "data": [
                  "806A74400188",
                  "1C",
                  "*DONE"
                ]
              }
            },
            "info": {
              "vcc": 3.24,
              "lastRestart": {
                "name": "P",
                "desc": "Powered Off"
              },
              "device": "ORP",
              "firmware": "2.13"
            },
            "address": 98,
            "typeId": 150,
            "sampling": 1,
            "isActive": false,
            "feeds": [],
            "triggers": [],
            "name": "ORP"
          },
          {
            "id": 2,
            "values": {
              "temperature": 22,
              "tempUnits": "C",
              "pH": 5.849
            },
            "options": {
              "name": "PH",
              "isProtocolLocked": false,
              "ledEnabled": true,
              "extendedScale": false,
              "calibrationMode": 3,
              "slope": {
                "acid": 93.5,
                "base": 91.1,
                "mV": 80.6
              },
              "readInterval": 15000,
              "calibration": {
                "len": 4,
                "total": 46,
                "data": [
                  "6C9D47CE34C3",
                  "6E39C3010101",
                  "00804000E040",
                  "002041894C",
                  "*DONE"
                ]
              }
            },
            "info": {
              "device": "pH",
              "firmware": "2.13",
              "vcc": 3.29,
              "lastRestart": {
                "name": "P",
                "desc": "Powered Off"
              }
            },
            "address": 99,
            "typeId": 100,
            "sampling": 1,
            "isActive": false,
            "feeds": [],
            "triggers": [],
            "name": "PH"
          },
          {
            "id": 3,
            "values": {
              "relays": [
                {
                  "id": 1,
                  "name": "Relay #1",
                  "enabled": false
                },
                {
                  "id": 2,
                  "name": "Relay #2",
                  "enabled": false
                },
                {
                  "id": 3,
                  "name": "Relay #3",
                  "enabled": false
                },
                {
                  "id": 4,
                  "name": "Relay #4",
                  "enabled": false
                },
                {
                  "id": 5,
                  "name": "Relay #5",
                  "enabled": false
                },
                {
                  "id": 6,
                  "name": "Relay #6",
                  "enabled": false
                },
                {
                  "id": 7,
                  "name": "Relay #7",
                  "enabled": false
                },
                {
                  "id": 8,
                  "name": "Relay #8",
                  "enabled": false
                }
              ]
            },
            "options": {
              "name": "Multi-Relay Board",
              "idType": "bit",
              "controllerType": "mcp23008",
              "maxRelays": 8
            },
            "info": {
              "registers": [
                {
                  "name": "IODIR",
                  "register": 0,
                  "desc": "I/O directions for 1-8"
                },
                {
                  "name": "IOPOL",
                  "register": 1,
                  "desc": "Input polarity for 1-8"
                },
                {
                  "name": "GPINTEN",
                  "register": 2,
                  "desc": "Interrupt on change for 1-8"
                },
                {
                  "name": "DEFVAL",
                  "register": 3,
                  "desc": "Default interrupt value for 1-8"
                },
                {
                  "name": "INTCON",
                  "register": 4,
                  "desc": "Interrupt control register for 1-8.  If 0 then interrupt fired on change else it is fired when matche$
                },
                {
                  "name": "IOCON",
                  "register": 5,
                  "desc": "I/O control per mcp23008 datasheet.  This determines whether in byte mode or sequential mode"
                },
                {
                  "name": "GPPU",
                  "register": 6,
                  "desc": "Pull up resistor for 1-8"
                },
                {
                  "name": "INTF",
                  "register": 7,
                  "desc": "Interrupt condition for 1-8"
                },
                {
                  "name": "INTCAP",
                  "register": 8,
                  "desc": "Interrupt capture at the time the interrupt occurred for 1-8"
                },
                {
                  "name": "OLAT",
                  "register": 10,
                  "desc": "Output latches for 1-8"
                }
              ]
            },
            "address": 19,
            "typeId": 300,
            "sampling": 1,
            "isActive": false,
            "feeds": [],
            "triggers": [],
            "name": "Multi-Relay Board"
          }
        ],
        "busNumber": 1,
        "functions": {
          "i2c": true,
          "tenBitAddr": false,
          "protocolMangling": false,
          "smbusPec": true,
          "smbusBlockProcCall": false,
          "smbusQuick": true,
          "smbusReceiveByte": true,
          "smbusSendByte": true,
          "smbusReadByte": true,
          "smbusWriteByte": true,
          "smbusReadWord": true,
          "smbusWriteWord": true,
          "smbusProcCall": true,
          "smbusReadBlock": false,
          "smbusWriteBlock": true,
          "smbusReadI2cBlock": true,
          "smbusWriteI2cBlock": true
        },
        "addresses": [
          {
            "address": 19,
            "manufacturer": 0,
            "product": 0,
            "name": "Multi-Relay Board"
          },
          {
            "address": 77,
            "manufacturer": 0,
            "product": 0,
            "name": "Unknown"
          },
          {
            "address": 98,
            "manufacturer": 0,
            "product": 0,
            "name": "ORP"
          },
          {
            "address": 99,
            "manufacturer": 0,
            "product": 0,
            "name": "PH"
          }
        ]
      }
    ],
    "detected": [
      {
        "driver": "bcm2835 (i2c@7e804000)",
        "path": "/sys/class/i2c-dev/i2c-1",
        "name": "i2c-1",
        "busNumber": 1
      }
    ]
  },
  "connections": [
    {
      "id": 2,
      "isActive": true,
      "options": {},
      "name": "njsPC - automatic",
      "type": "njspc",
      "protocol": "http:",
      "ipAddress": "192.168.0.20",
      "port": 4200,
      "userName": "",
      "password": "",
      "sslKeyFile": "",
      "sslCertFile": ""
    }
  ],
  "genericDevices": {
    "isActive": true,
    "buses": [],
    "detected": [],
    "options": {},
    "devices": [
      {
        "id": 1,
        "values": {
          "units": "C",
          "temp": 52.616,
          "": 123.8036
        },
        "options": {
          "name": "Pi Temp",
          "readInterval": 3000,
          "units": "C"
        },
        "info": {},
        "triggers": [],
        "feeds": [],
        "typeId": 1101,
        "isActive": true,
        "name": "Pi Temp"
      },
      {
        "id": 2,
        "values": {
          "isOn": true,
          "psi": null,
          "latch": 10000,
          "voltage": null
        },
        "options": {
          "name": "Filter Pressure",
          "inducerOffset": 0.5,
          "psiPerVolt": 25,
          "alarm": {
            "enable": false,
            "pressure": 0,
            "units": {
              "name": ""
            },
            "tolerance": 0
          }
        },
        "info": {},
        "triggers": [],
        "feeds": [
          {
            "id": 2,
            "isActive": true,
            "changesOnly": false,
            "sampling": 1,
            "connectionId": -1,
            "deviceBinding": "generic:202:2",
            "sendValue": "psi",
            "property": "voltage",
            "payloadExpression": ""
          }
        ],
        "typeId": 202,
        "isActive": true,
        "name": "Filter Pressure"
      }
    ]
  },
  "oneWire": {
    "isActive": false,
    "buses": [
      {
        "id": 1,
        "isActive": true,
        "devices": [
          {
            "id": 1,
            "values": {
              "units": "F",
              "temp": 80.48660000000001
            },
            "options": {
              "readInterval": 3000,
              "name": "Water Temp",
              "units": "F",
              "calibration": 0,
              "alarmLow": 158,
              "alarmHigh": 167,
              "resolution": 12
            },
            "info": {
              "convTime": 750,
              "extPower": "External Power"
            },
            "address": "28-00000c8716df",
            "typeId": 6,
            "sampling": 1,
            "isActive": true,
            "feeds": [
              {
                "id": 1,
                "isActive": true,
                "changesOnly": false,
                "sampling": 1,
                "connectionId": 1,
                "sendValue": "tempF",
                "eventName": "temps",
                "property": "waterSensor1"
              }
            ],
            "triggers": [],
            "name": "Water Temp"
          },
          {
            "id": 2,
            "values": {
              "units": "F",
              "temp": 78.6866
            },
            "options": {
              "readInterval": 3000,
              "name": "Air Temp",
              "units": "F",
              "calibration": 0,
              "alarmLow": 41,
              "alarmHigh": 185,
              "resolution": 12
            },
            "info": {
              "convTime": 750,
              "extPower": "External Power"
            },
            "address": "28-3de10457fd0e",
            "typeId": 6,
            "sampling": 1,
            "isActive": true,
            "feeds": [],
            "triggers": [],
            "name": "Air Temp"
          }
        ],
        "busNumber": 1,
        "addresses": [
          {
            "address": "28-3de10457fd0e",
            "manufacturer": 0,
            "product": 0,
            "name": "28-3de10457fd0e"
          },
          {
            "address": "28-00000c8716df",
            "manufacturer": 0,
            "product": 0,
            "name": "28-00000c8716df"
          }
        ]
      }
    ],
    "detected": [
      {
        "path": "/sys/devices/w1_bus_master1",
        "name": "w1_bus_master1",
        "busNumber": 1
      }
    ]
  },
  "lastUpdated": "5/18/2023, 9:23:54 PM",
  "configVersion": 1
}

**Reply by rstrouse** [reply]
It looks like we got one of the generic device definitions mixed up in the spi device list.  There were two Pressure Transducer items in the dropdown.  Until I get the definitions reworked you will need to stop REM then edit the controllerConfig.json file.  You will need to change the deviceId value from 202 to 200 for SPI0 channel 2.  It should never have been marked as an spi available device type.

**Reply by thrystan** [reply]
Ok, that worked, thanks.
Is there any documentation for adding a sensor type? One of the sensors I'm using is an eTape liquid level sensor. https://milonetech.com/products/standard-etape-assembly and I'm interested in adding it.

**Reply by rstrouse** [reply]
I am giving the SPI definitions a bit of an overhaul.  They haven't had the attention that the I2c devices have had.  This should simplify the configuration for these devices.

**Reply by rstrouse** [reply]
If you pull REM I have improved the definition of such devices as pressure transducers and NTC temp probes on the SPI bus.  You can now defined these directly with SPI and output their values with a feed to anywhere you desire.

**Reply by thrystan** [reply]
Is this what I should be seeing?
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/3c91677f-ccfa-4b30-8c62-edd6d9ec26a0)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/debbb9ea-aa10-4715-828f-360b009c9bea)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/ee76f624-9c92-4250-b55c-31ee3f588d87)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/1f9ce508-9512-4156-85bf-eda1ec2cf5ce)

**Reply by rstrouse** [reply]
Hmmm.  Are you getting readings from the SPI chip.  The raw values should be appearing on the channel bar.  Are you witnessing any errors in the logs.  You also do not need to send raw values to the internal device as it will do all the pressure conversions without it.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/3f976a8a-2021-4c40-81e6-432107fc0385)

**Reply by rstrouse** [reply]
Hard refresh your browser.  It looks like the spi.js file is out of date in the cache.  Setting up the feed should look like this.  And it should be the only feed you need.  <-- See what I did there.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/60b2d90f-6455-4dce-a836-260fd914edf4)

**Comment by rstrouse** [comment]
I’ll have a look

**Comment by thrystan** [comment]
I do not see that.

**Comment by thrystan** [comment]
I am not seeing readings.

**Comment by rstrouse** [comment]
Did you clear the cache in the browser?

**Comment by thrystan** [comment]
yes, and nothing changed.

**Reply by rstrouse** [reply]
Are you seeing any browser errors?

**Reply by rstrouse** [reply]
Also do you have the SPI interface enabled on the Raspberry Pi under raspi config.  Even without an MCP chip connected I get a reading of 0/0 and no reading when the interface is disabled.

![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/421b7ef4-be8b-4c4f-b6b7-8a9fd6959aa9)

**Comment by thrystan** [comment]
Yes, I can read it from a python script using the adafruit library, so I
know the hardware is working.

**Reply by rstrouse** [reply]
Ok so I am a bit confused.  Please make sure you have the latest pull and lets watch the logs.  I added a number of outputs to the logs when the logging level is set to be verbose.

To do this follow the steps below.
1. Go into the General tab and change the logging level to `Verbose`.  Then click the `Log to File` checkbox.
2. Restart REM either using `pm2 restart REM` if you are using pm2 or reboot the pi if you are using sys.
3. Wait until REM comes back up completely then git it another minute.
4. Go back to the General tab and change the logging level back to `Info` and un-check the `Log to File` checkbox.
5. REM will have written a file with today's date to the `~/relayEquipmentManager/logs` folder.  Please upload that file here.

**Reply by thrystan** [reply]
[consoleLog(2023-05-25_11-20-01).log](https://github.com/tagyoureit/nodejs-poolController/files/11568241/consoleLog.2023-05-25_11-20-01.log)
This has a couple of false starts, because the controllerConfig.json had a bad spi config. I fixed that then it started.

**Comment by rstrouse** [comment]
What is trying to set an SPI device to true?  Do you have something set up to latch an spi device as if it were a relay.  Like a circuit definition or something?  The mcp3008 chip only takes analog inputs and converts them to a digital signal.  There is no need to write to that device at all.

```
[5/25/2023, 11:20:54 AM] verbose: 192.168.0.20 PUT /state/device/spi:0:2 {"isOn":true,"latch":10000}
[5/25/2023, 11:20:54 AM] info: [11:20:54 AM] 192.168.0.20 PUT /state/device/spi:0:2 {"isOn":true,"latch":10000}
```

**Comment by thrystan** [comment]
I have no idea, lol. I tried setting it up in the SPI tab.

**Reply by thrystan** [reply]
It feels like it isn't reading the chip, which is why I was asking about which gpio pin it was expecting to use as the cs line. I'm using the CE0/gpio8 pin.

**Reply by rstrouse** [reply]
I expected to see logs like below if you had Verbose logging turned on.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/54518db6-0ec5-410a-a8c6-230ad5463a4b)



Here is a chart.  And the second image is cooked up from the web for SPI0.  The CS pin will enable the chip as soon as the interface is opened.  You should be seeing 
![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/2e381373-90d6-4ed6-b564-1d9b0418a70a)

![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/67099beb-2680-4802-b945-9d3392f3dc9b)

**Reply by thrystan** [reply]
That is how I have it hooked up. when I look at the verbose logging in monit I see this.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/8f84bd39-5116-4a4f-af56-b09a8eb943b8)
[consoleLog(2023-05-25_17-38-06).log](https://github.com/tagyoureit/nodejs-poolController/files/11570282/consoleLog.2023-05-25_17-38-06.log)
Stuff from the 1-wire sensors, but nothing from the mcp3008.

**Reply by rstrouse** [reply]
Lets just make sure you are running the current software.  Run the following commands and post the results of the git pull command here.  Something isn't right.  Either the channel isn't enabled (but I saw it green earlier), the bus is not initializing (but I saw it in your previous log), or the software is not up to date.  Also you should turn off SPI1 if you don't have any endpoints on it.
`cd ~/relayEquipmentManager`
`git pull`
`npm i`

**Reply by tagyoureit** [reply]
Also, do not run your Python script at the same time.  That could cause access issues.

**Reply by thrystan** [reply]
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/67a2402b-33a3-49e5-bfa6-61cc95946e18)
I've turned off my scripts and disabled spi1, I had turned it on just to make sure I wasn't checking the wrong bus.

I'm still not seeing what you had posted above, and when I turn on the channel and save it REM errors out when it restarts.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/2dbe8861-a2b2-4004-95be-b10ccd808f31)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/73d043bc-1362-46a4-b42b-2efff56e6a70)

**Reply by rstrouse** [reply]
What is the error that it is reporting when it restarts?

**Reply by thrystan** [reply]
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/f9722ce6-e16b-4a15-97d9-6a72c1c02100)

**Reply by thrystan** [reply]
The issue seems to be from here. 
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/f0a8348d-18fe-4b5e-b4a4-b72d54369c02)

**Reply by rstrouse** [reply]
Again this looks like you are not running the current code.  The device reference was added on this latest round of updates.  If you do a git pull and it comes back as being up to date.  Do the following:

`pm2 stop REM`
`cd ~/relayEquipmentManager`
`npm run build`
`pm2 restart REM`

**Reply by thrystan** [reply]
Ok, I did that and I'm getting some differences. But I keep getting:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/da2e10f7-51c6-4e78-b92b-f066b96e9992)
And when I try to setup a feed, I get this:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/e7ba8220-1d29-4ce3-bc8d-77bad84228ea)

**Reply by thrystan** [reply]
[consoleLog(2023-05-26_14-04-32).log](https://github.com/tagyoureit/nodejs-poolController/files/11579952/consoleLog.2023-05-26_14-04-32.log)
[consoleLog(2023-05-26_14-05-39).log](https://github.com/tagyoureit/nodejs-poolController/files/11579953/consoleLog.2023-05-26_14-05-39.log)
[consoleLog(2023-05-26_14-06-18).log](https://github.com/tagyoureit/nodejs-poolController/files/11579955/consoleLog.2023-05-26_14-06-18.log)
Here are some verbose logs.

**Comment by rstrouse** [comment]
That is much better.  It sounds like you removed the `npm run build` from the pm2 script.  That means you must perform that command each time you do a `git pull`.   We have been chasing the same issue without resolution because the typescript compiler never ran to build the new code into your deployment.

Now you need to go into your njsPC implementation and find out which circuit you have set up to write to SPI channel 2.  These cannot be written to they are read only.  That is what the first screenshot is indicating.


As for the unhandled promise I have made a code change to catch that promise.  This occurs because the read was occurring while the chip was being reset.  Please pull REM and don't forget to do `npm run build`.

**Reply by thrystan** [reply]
Yes, I had to remove the build because on the zero 2 I have to modify the max memory 'export NODE_OPTIONS=--max_old_space_size=1024' or it errors out after running out of memory.

I found the circuit that was doing the writing, so that is resolved now.

When I add a new channel I am getting this
![image](https://github.com/tagyoureit/nodejs-poolController/assets/8936451/36333ef7-eb78-4538-88c7-f914359fecfb)

And I'm still not getting any data.

**Reply by rstrouse** [reply]
Make sure you hard refresh your browser now that it is serving the proper files.  That ENOENT is odd considering how it actually just opened the file in the line above.

**Reply by rstrouse** [reply]
When you switch to verbose logging do you see Reading SPI0 Channel 2?

**Comment by thrystan** [comment]
No, just channel 0 and channel 1. I appear to sometimes get good data from
channel 0. but not from channel 1

**Reply by rstrouse** [reply]
Is it reading 0 or garbage

**Comment by thrystan** [comment]
0,
Also somehow the configuration got lost and now it's not finding the njsPC
server and when I manually added it it's getting timeout errors.

**Reply by rstrouse** [reply]
Timeout errors simple mean that it cannot find it.  Are you using the loopback or are you using its assigned IP address?

**Comment by thrystan** [comment]
I tried both.
